generate_brick.py

```diff
@@ -108,42 +124,28 @@ def add_class_restriction(klass, definition):
     G.add( (equivalent_class, OWL.intersectionOf, list_name) )
     c = Collection(G, list_name, l)
 
-def define_subclasses(definitions, superclass):
-    for subclass, properties in definitions.items():
-        G.add((superclass, A, OWL.Class))
-        G.add((BRICK[subclass], A, OWL.Class))
-        G.add((BRICK[subclass], RDFS.label, Literal(subclass.replace("_"," "))))
-        G.add((BRICK[subclass], RDFS.subClassOf, superclass))
-        for k, v in properties.items():
-            if isinstance(v, list) and k == "tags":
-                add_tags(subclass, v)
-            elif isinstance(v, list) and k == "parents":
-                for parent in v:
-                    G.add( (BRICK[subclass], RDFS.subClassOf, parent) )
-            elif isinstance(v, list) and k == "substances":
-                add_restriction(subclass, v)
-            elif not apply_prop(subclass, k, v):
-                if isinstance(v, dict) and k == "subclasses":
-                    define_subclasses(v, BRICK[subclass])
 
-def define_measurable_subclasses(definitions, measurable_class):
+def define_subclasses(definitions, superclass, punning_level=0):
+    G.add((superclass, A, OWL.Class))
     for subclass, properties in definitions.items():
         G.add((BRICK[subclass], A, OWL.Class))
-        G.add((BRICK[subclass], A, BRICK[subclass]))
-        G.add((BRICK[subclass], RDFS.label, Literal(subclass.replace("_"," "))))
-        # first level: we are instances of the measurable_class
-        G.add((BRICK[subclass], A, measurable_class))
+        G.add((BRICK[subclass], RDFS.label, Literal(subclass.replace("_", " "))))
+        if punning_level >= 1:
```

Comment:
I'm *very* confused what you mean by "punning" here. This is not the common definition of "punning" as it is used in OWL 2 modeling. Maybe this PR can stick to checking whether or not tags are applied and not touch the logic for how Brick gets generated?

Reply:
There are a couple of problems in the previous code:
1. `add_tags` is not called if there is no tag defined. Thus, it's hard to put the check logic.
2. The parsing logic of `bricksrc/*.py` is implemented in two different places, so it's hard to maintain the readability and the sanity of the logic.
3. The logic `apply_prop` is not straightforward as it's not just applying remaining properties but also used to check if it needs to add subclasses. They should be separated out.

The current implementation make the logic more readable and unifies the parsing logic for substances/properties and the others though there are some rooms for improvement. I can do as you suggested, but, then, I have to duplicate the logic in different places, which I commonly avoid.

I defined `punning_level` to indicate the level of hierarchy you want to associate with the class.
- `Temperature a Temperature` -> Level 1
- `Temperature a Quantity` -> Level 2 (based on the hierarchy)
I can use a different variable name if you have other suggestions.

A different note (may need a different thread): what is the benefit of using `a` for hierarchical relations other than subclasses? I thought a class being an instance of itself was enough for our use cases. What's the benefit of the above design compared with this:
- `Temperature a Temperature`
- `Temperature rdfs:subClassOf Quantity`

Reply:
I think the proposed implementation complicates the generation of the ontology, and confuses the issues of what is explicitly communicated by the ontology vs what is implied by the model. Please allow me to file my own PR on refactoring the logic to make it more readable